File: src/Mobile/ntk_note_ip_app/lib/core/history/ip_history_sync_service.dart
import '../../domain/entities/ip_lookup_record.dart';
import '../../domain/usecases/action_lookup_ip.dart';
import '../../domain/usecases/get_list_ip_lookup_records.dart';
import 'ip_history_store.dart';

class IpHistorySyncService {
  IpHistorySyncService(
    this._historyStore,
    this._getListRecords,
    this._actionLookup,
  );

  final IpHistoryStore _historyStore;
  final GetListIpLookupRecordsUseCase _getListRecords;
  final ActionLookupIpUseCase _actionLookup;

  bool _syncedThisSession = false;

  void resetSession() {
    _syncedThisSession = false;
  }

  Future<void> syncAfterAuthentication() async {
    if (_syncedThisSession) {
      return;
    }

    final serverResult = await _getListRecords();
    if (!serverResult.isSuccess || serverResult.data == null) {
      return;
    }

    final serverRecords = serverResult.data!;
    await _mergeServerIntoLocal(serverRecords);

    final serverAddresses = serverRecords
        .map((r) => r.address.toLowerCase())
        .toSet();
    final local = await _historyStore.getEntries();
    final toUpload = local
        .where((e) => !serverAddresses.contains(e.address.toLowerCase()))
        .toList();

    for (final entry in toUpload) {
      await _actionLookup(entry.address);
    }

    _syncedThisSession = true;
  }

  Future<void> _mergeServerIntoLocal(List<IpLookupRecord> serverRecords) async {
    for (final record in serverRecords) {
      await _historyStore.add(
        address: record.address,
        isIPv6: record.address.contains(':'),
        city: record.city,
        countryCode: record.countryCode,
        recordedAt: record.created,
      );
    }
  }
}

File: src/Web/ClientApp/src/app/core/ip-history-sync.service.ts
import { Injectable } from '@angular/core';
import { catchError, forkJoin, map, Observable, of, switchMap } from 'rxjs';
import { IpHistoryService } from './ip-history.service';
import { IpLookupRecordDto, IpLookupService } from '../ip-lookup/ip-lookup.service';

@Injectable({ providedIn: 'root' })
export class IpHistorySyncService {
  private syncedThisSession = false;

  constructor(
    private readonly ipLookup: IpLookupService,
    private readonly ipHistory: IpHistoryService
  ) {}

  syncAfterAuthentication(): Observable<void> {
    if (this.syncedThisSession) {
      return of(void 0);
    }

    return this.ipLookup.getListIpLookupRecords().pipe(
      switchMap((serverRecords) => {
        this.mergeServerIntoLocal(serverRecords);
        const serverAddresses = new Set(
          serverRecords.map((r) => r.address.toLowerCase())
        );
        const toUpload = this.ipHistory
          .getEntries()
          .filter((entry) => !serverAddresses.has(entry.address.toLowerCase()));

        if (toUpload.length === 0) {
          this.syncedThisSession = true;
          return of(void 0);
        }

        return forkJoin(
          toUpload.map((entry) =>
            this.ipLookup.actionLookup(entry.address).pipe(catchError(() => of(null)))
          )
        ).pipe(map(() => void 0));
      }),
      map(() => {
        this.syncedThisSession = true;
      }),
      catchError(() => of(void 0))
    );
  }

  resetSession(): void {
    this.syncedThisSession = false;
  }

  private mergeServerIntoLocal(serverRecords: IpLookupRecordDto[]): void {
    for (const record of serverRecords) {
      this.ipHistory.add({
        address: record.address,
        isIPv6: record.address.includes(':'),
        city: record.city ?? undefined,
        countryCode: record.countryCode ?? undefined,
        recordedAt: record.created,
      });
    }
  }
}
